torch/distributed/fsdp/_optim_utils.py

```diff
@@ -155,6 +155,7 @@ def _communicate_optim_state(
                 buffer_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
                 tensor_buffer = value.new_zeros(*buffer_size)
             dist._all_gather_base(tensor_buffer, value, group=group)
+            torch.cuda.synchronize()
```

Comment:
does a barrier() fix it as well?

Reply:
I tried with
```
dist.barrier(group=group)
```
and the hang is fixed as well.